sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathfunctions.scala

```diff
@@ -0,0 +1,272 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.expressions
+
+import org.apache.spark.sql.catalyst.analysis.UnresolvedException
+import org.apache.spark.sql.types._
+
+trait MathematicalExpression extends UnaryExpression with Serializable { self: Product =>
+  type EvaluatedType = Any
+
+  override def dataType: DataType = DoubleType
+  override def foldable: Boolean = child.foldable
+  override def nullable: Boolean = true
+
+  lazy val numeric = child.dataType match {
+    case n: NumericType => n.numeric.asInstanceOf[Numeric[Any]]
+    case other => sys.error(s"Type $other does not support numeric operations")
+  }
+}
+
+abstract class MathematicalExpressionForDouble(f: Double => Double)
+  extends MathematicalExpression { self: Product =>
+
+  override def eval(input: Row): Any = {
+    val evalE = child.eval(input)
+    if (evalE == null) {
+      null
+    } else {
+      f(numeric.toDouble(evalE))
+    }
+  }
+}
+
+abstract class MathematicalExpressionForInt(f: Int => Int)
+  extends MathematicalExpression { self: Product =>
+
+  override def dataType: DataType = IntegerType
+
+  override def eval(input: Row): Any = {
+    val evalE = child.eval(input)
+    if (evalE == null) {
+      null
+    } else {
+      f(numeric.toInt(evalE))
+    }
+  }
+}
+
+abstract class MathematicalExpressionForFloat(f: Float => Float)
+  extends MathematicalExpression { self: Product =>
+
+  override def dataType: DataType = FloatType
+
+  override def eval(input: Row): Any = {
+    val evalE = child.eval(input)
+    if (evalE == null) {
+      null
+    } else {
+      f(numeric.toFloat(evalE))
+    }
+  }
+}
+
+abstract class MathematicalExpressionForLong(f: Long => Long)
+  extends MathematicalExpression { self: Product =>
+
+  override def dataType: DataType = LongType
+
+  override def eval(input: Row): Any = {
+    val evalE = child.eval(input)
+    if (evalE == null) {
+      null
+    } else {
+      f(numeric.toLong(evalE))
+    }
+  }
+}
+
+case class Sin(child: Expression) extends MathematicalExpressionForDouble(math.sin) {
+  override def toString: String = s"SIN($child)"
```

Comment:
This is a style nit so feel free to ignore, but you might just make the name an arg to the constructor so you don't have to write this method each time.  Similar to how we handle `BinaryExpression`.